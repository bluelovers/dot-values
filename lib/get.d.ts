/**
 * @param  {Object} target
 * @param  {String} path
 * @param  {*}      value
 * @return {*}
 */
export declare function get<T, D = T>(target: object, path: string | string[], defaultValue?: D): T | D | (T | D)[];
export default get;
