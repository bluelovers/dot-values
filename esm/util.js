/// <reference path="../module.d.ts" />
import _get from 'get-value';
// @ts-ignore
import escapeSplit from 'escape-split';
const SymStar = Symbol('*');
const splitSpecific = escapeSplit({ delimiter: '.', escaper: '\\' });
export function r(fn, target, paths, extra, result) {
    paths = p(paths);
    return r0(fn, target, paths, extra, result);
}
export function r0(fn, target, paths, extra, result) {
    let symIndex = paths.indexOf(SymStar), hasResult = result != null, isLast = !paths.length;
    if (symIndex < 0) {
        let value = isLast
            ? target
            // @ts-ignore
            : fn(target, paths, extra);
        if (hasResult) {
            result.push(value);
        }
        return value;
    }
    result = result || [];
    if (symIndex === 0) {
        let keys = ObjectKeysExtra(target, {
            //nonEnumerable: true,
            symbol: true,
        });
        paths.shift();
        keys
            .forEach(k => {
            r0(fn, target, [k].concat(paths), extra, result);
        });
        return result;
    }
    else {
        let p1 = paths.slice(0, symIndex);
        let p2 = paths.slice(symIndex);
        return r0(fn, _get(target, p1), p2, extra, result);
    }
}
export function ObjectKeysExtra(obj, options = {}) {
    let keys = options.nonEnumerable ?
        Object.getOwnPropertyNames(obj)
        : Object.keys(obj);
    if (options.symbol) {
        keys.push(...Object.getOwnPropertySymbols(obj));
    }
    return keys;
}
/**
 * handle input to paths
 *
 * @param {string | ITSKeys[]} paths
 * @returns {ITSKeys[]}
 */
export function p(paths) {
    return p1(p0(paths));
}
export function p0(paths) {
    if (Array.isArray(paths))
        return paths.slice();
    return splitSpecific(paths);
}
export function p1(paths) {
    return paths
        .map(v => {
        if (v === '*') {
            return SymStar;
        }
        else if (v === '\\*') {
            v = '*';
        }
        return v;
    });
}
/**
 * check all value is true
 *
 * @param {boolean | boolean[]} result
 * @returns {boolean}
 */
export function b(result) {
    return Array.isArray(result)
        ? result.every(boolean => !!boolean)
        : result;
}
export function b_some(result) {
    return Array.isArray(result)
        ? result.some(boolean => !!boolean)
        : result;
}
//# sourceMappingURL=util.js.map