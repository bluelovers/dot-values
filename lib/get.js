"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
const get_value_1 = __importDefault(require("get-value"));
function get(target, paths, defaultValue) {
    return util_1.r(get_value_1.default, target, paths, { default: defaultValue });
}
exports.get = get;
exports.default = get;
//# sourceMappingURL=get.js.map