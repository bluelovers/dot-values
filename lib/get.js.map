{"version":3,"file":"get.js","sourceRoot":"","sources":["get.ts"],"names":[],"mappings":";;AAAA,iCAA8B;AAC9B,yCAA6B;AAE7B;;;;;GAKG;AACH,SAAgB,GAAG,CAAW,MAAc,EAAE,IAAuB,EAAE,YAAgB;IAEtF,OAAO,QAAC,CAAQ,mBAAI,EAAE,MAAM,EAAE,QAAC,CAAC,IAAI,CAAC,EAAE,EAAE,OAAO,EAAE,YAAY,EAAE,CAAC,CAAA;AAClE,CAAC;AAHD,kBAGC;AAED,kBAAe,GAAG,CAAA","sourcesContent":["import { r, p } from './util';\nimport _get from 'get-value';\n\n/**\n * @param  {Object} target\n * @param  {String} path\n * @param  {*}      value\n * @return {*}\n */\nexport function get<T, D = T>(target: object, path: string | string[], defaultValue?: D)\n{\n\treturn r<D | T>(_get, target, p(path), { default: defaultValue })\n}\n\nexport default get\n"]}