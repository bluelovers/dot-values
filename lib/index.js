"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const get_1 = __importDefault(require("./get"));
exports.get = get_1.default;
exports.getValue = get_1.default;
const set_1 = __importDefault(require("./set"));
exports.set = set_1.default;
exports.setValue = set_1.default;
const has_1 = __importDefault(require("./has"));
exports.has = has_1.default;
exports.hasValue = has_1.default;
const unset_1 = __importDefault(require("./unset"));
exports.unset = unset_1.default;
exports.unsetValue = unset_1.default;
exports.default = {
    get: get_1.default,
    set: set_1.default,
    has: has_1.default,
    unset: unset_1.default,
};
//# sourceMappingURL=index.js.map