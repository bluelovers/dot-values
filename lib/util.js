"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="../module.d.ts" />
const get_value_1 = __importDefault(require("get-value"));
// @ts-ignore
const escape_split_1 = __importDefault(require("escape-split"));
const SymStar = Symbol('*');
const splitSpecific = escape_split_1.default({ delimiter: '.', escaper: '\\' });
function r(fn, target, paths, extra, result) {
    paths = p(paths);
    return r0(fn, target, paths, extra, result);
}
exports.r = r;
function r0(fn, target, paths, extra, result) {
    let symIndex = paths.indexOf(SymStar), hasResult = result != null, isLast = !paths.length;
    if (symIndex < 0) {
        let value = isLast
            ? target
            // @ts-ignore
            : fn(target, paths, extra);
        if (hasResult) {
            result.push(value);
        }
        return value;
    }
    result = result || [];
    if (symIndex === 0) {
        let keys = ObjectKeysExtra(target, {
            //nonEnumerable: true,
            symbol: true,
        });
        paths.shift();
        keys
            .forEach(k => {
            r0(fn, target, [k].concat(paths), extra, result);
        });
        return result;
    }
    else {
        let p1 = paths.slice(0, symIndex);
        let p2 = paths.slice(symIndex);
        return r0(fn, get_value_1.default(target, p1), p2, extra, result);
    }
}
exports.r0 = r0;
function ObjectKeysExtra(obj, options = {}) {
    let keys = options.nonEnumerable ?
        Object.getOwnPropertyNames(obj)
        : Object.keys(obj);
    if (options.symbol) {
        keys.push(...Object.getOwnPropertySymbols(obj));
    }
    return keys;
}
exports.ObjectKeysExtra = ObjectKeysExtra;
/**
 * handle input to paths
 *
 * @param {string | ITSKeys[]} paths
 * @returns {ITSKeys[]}
 */
function p(paths) {
    return p1(p0(paths));
}
exports.p = p;
function p0(paths) {
    if (Array.isArray(paths))
        return paths.slice();
    return splitSpecific(paths);
}
exports.p0 = p0;
function p1(paths) {
    return paths
        .map(v => {
        if (v === '*') {
            return SymStar;
        }
        else if (v === '\\*') {
            v = '*';
        }
        return v;
    });
}
exports.p1 = p1;
/**
 * check all value is true
 *
 * @param {boolean | boolean[]} result
 * @returns {boolean}
 */
function b(result) {
    return Array.isArray(result)
        ? result.every(boolean => !!boolean)
        : result;
}
exports.b = b;
function b_some(result) {
    return Array.isArray(result)
        ? result.some(boolean => !!boolean)
        : result;
}
exports.b_some = b_some;
//# sourceMappingURL=util.js.map