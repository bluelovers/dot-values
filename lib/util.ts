import _get from 'get-value';
import _set from 'set-value';
import _has from 'has-value';
import _unset from 'unset-value';

/**
 * @param  {String} fn
 * @param  {Object} target
 * @param  {Array.<String>} path
 * @param  {*} extra
 * @param  {Array} result (optional)
 * @return {*}
 */
export function r<T>(fn: typeof _get | typeof _set | typeof _has | typeof _unset, target: unknown | unknown[], path: string[], extra?: unknown, result?: T[]): T | T[]
{
	let symIndex = path.indexOf('*'),
		hasResult = Array.isArray(result),
		isLast = !path.length

	if (symIndex < 0 || isLast)
	{
		let value = isLast
			? target
			// @ts-ignore
			: fn(target, path.join('.'), extra)

		if (hasResult)
		{
			result.push(value)
		}

		return value
	}

	result = hasResult ? result : []

	if (!(symIndex === 0 && Array.isArray(target)))
	{
		return r(fn, _get(target as unknown[], path.shift()), path, extra, result)
	}

	path.shift()
	target.forEach((target) =>
	{
		r(fn, target, Array.from(path), extra, result)
	})

	return result
}

/**
 * @param {String|Array} path
 * @return {Array.<String>}
 */
export function p(path: string | string[])
{
	if (Array.isArray(path)) return path

	return path.split(/\.*(\*)\.*/)
		.filter(segment => segment !== '')
}

/**
 * @param {Array|Boolean} result
 * @return {Boolean}
 */
export function b(result: boolean | boolean[])
{
	return Array.isArray(result)
		? !result.filter(boolean => !boolean).length
		: result
}
