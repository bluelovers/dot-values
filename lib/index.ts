import get from './get';
import set from './set';
import has from './has';
import unset from './unset';

export {
	get,
	set,
	has,
	unset,
}

export {
	get as getValue,
	set as setValue,
	has as hasValue,
	unset as unsetValue,
}

export default {
	get,
	set,
	has,
	unset,
}
