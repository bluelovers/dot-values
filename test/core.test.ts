/**
 * Created by User on 2020/1/12.
 */

import _get, { Options } from 'get-value';
import _set from 'set-value';
import _has from 'has-value';
import _unset from "../lib/core/unset";
import { p, b } from '../lib/util';

let testData = JSON.parse(JSON.stringify(require('./deep.json'))) as typeof import('./deep.json');

// @ts-ignore
describe(`check core function support input array paths`, () =>
{
	it(`get`, function ()
	{

		let actual = _get(testData, p('a.b.c.arr.0') as any);
		let expected = {
			d: {
				e: { arr: [{ f: true }, { f: true }] },
			},
			x: true,
		};

		expect(actual).toStrictEqual(expected);
		expect(actual).toMatchSnapshot();
	});

	it(`set`, function ()
	{

		let expected = 999;

		// @ts-ignore
		_set(testData, p('a.b.c.arr.0'), expected)

		let actual = _get(testData, p('a.b.c.arr.0') as any);

		expect(actual).toStrictEqual(expected);
		expect(actual).toMatchSnapshot();
	});

	it(`has`, function ()
	{
		// @ts-ignore
		expect(_has(testData, p('a.b.c.arr.0'))).toBeTruthy();
		// @ts-ignore
		expect(_has(testData, p('a.b.c.arr.99'))).toBeFalsy();

	});

	it(`unset`, function ()
	{
		let paths = p('a.b.c.arr.2');

		expect(_has(testData, paths as any)).toBeTruthy();

		expect(_unset(testData, paths as any)).toBeTruthy();

		expect(_has(testData, paths as any)).toBeFalsy();

	});

})

// @ts-ignore
describe(`check core function support escape path`, () =>
{

	it(`get`, function ()
	{

		let actual = _get(testData, 'a.b\\.c.arr.0');
		let expected = {
			d: {
				e: { arr: [{ f: true }, { f: true }] },
			},
			y: true,
		};

		expect(actual).toStrictEqual(expected);
		expect(actual).toMatchSnapshot();
	});

	it(`set`, function ()
	{
		let paths = 'a.b\\.c.arr.0';

		let expected = 999;

		// @ts-ignore
		_set(testData, paths, expected)

		let actual = _get(testData, paths);

		expect(actual).toStrictEqual(expected);
		expect(actual).toMatchSnapshot();
	});

	it(`has`, function ()
	{
		// @ts-ignore
		expect(_has(testData, 'a.b\\.c.arr.0')).toBeTruthy();
		// @ts-ignore
		expect(_has(testData, 'a.b\\.c.arr.99')).toBeFalsy();

	});

	it(`unset`, function ()
	{
		let paths = 'a.b\\.c.arr.2';

		//console.dir(p(paths))

		expect(_has(testData, paths as any)).toBeTruthy();

		expect(_get(testData, paths as any)).toStrictEqual({
				d: {
					e: { arr: [{ f: false }, { f: false }] },
				},
				y: true,
			},
		);

		expect(_unset(testData, paths as any)).toBeTruthy();

		expect(_has(testData, paths as any)).toBeFalsy();

	});

})

