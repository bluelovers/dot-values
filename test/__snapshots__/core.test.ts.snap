// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check core function support escape path get 1`] = `
{
  "d": {
    "e": {
      "arr": [
        {
          "f": true,
        },
        {
          "f": true,
        },
      ],
    },
  },
  "y": true,
}
`;

exports[`check core function support escape path set 1`] = `999`;

exports[`check core function support input array paths get 1`] = `
{
  "d": {
    "e": {
      "arr": [
        {
          "f": true,
        },
        {
          "f": true,
        },
      ],
    },
  },
  "x": true,
}
`;

exports[`check core function support input array paths set 1`] = `999`;
